File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {    
public:
    int helper(TreeNode* cur, int sum, long long &ans){
        if (NULL == cur) return 0;
        int left = helper(cur->left, sum, ans);
        int right = helper(cur->right, sum, ans);
        int s = left + right + cur->val;
        long long product = 1ll * s * (sum - s);
        ans = max(ans, product);
        return s;
    }
    
    int maxProduct(TreeNode* root) {
        long long ans = 0;
        int total = helper(root, 0, ans);
        ans = 0;
        helper(root, total, ans);
        return ans % 1'000'000'007;
    }
};

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findSum(self, root):
        if not root:
            return 0
        left = self.findSum(root.left)
        right = self.findSum(root.right)
        return left + right + root.val
    
    def maxP(self, cur, mx, s):
        if not cur:
            return 0
        
        left = self.maxP(cur.left, mx, s)
        right = self.maxP(cur.right, mx, s)
        
        summ = left + right + cur.val
        mx[0] = max(summ *(s - summ), mx[0])
        return summ
    
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        summ = self.findSum(root)
        ans = [0]
        MOD = 1_000_000_007
        self.maxP(root, ans, summ)
        return ans[0] % MOD